openmls/src/extensions/mod.rs

```diff
@@ -166,6 +166,200 @@ pub enum Extension {
     ExternalSenders(ExternalSendersExtension),
 }
 
+/// A list of extensions with unique extension types.
+#[derive(
+    Debug, Clone, PartialEq, Eq, Serialize, Deserialize, TlsSize, TlsSerialize, TlsDeserialize,
+)]
+pub struct Extensions {
+    inner: Vec<Extension>,
+}
+
+impl Extensions {
+    /// Create an extension list that is empty.
+    pub fn empty() -> Self {
+        Self { inner: Vec::new() }
+    }
+
+    /// Create an extension list that contains a single extension.
+    pub fn single(extension: Extension) -> Self {
+        Self {
+            inner: vec![extension],
+        }
+    }
+
+    /// Create an extension list that contains multiple extensions.
+    ///
+    /// This function will fail when the list of extensions contains duplicate extension types.
+    pub fn multi(extensions: Vec<Extension>) -> Result<Self, InvalidExtensionError> {
+        extensions.try_into()
+    }
+
+    // ---------------------------------------------------------------------------------------------
+
+    /// Add an extension to the extension list.
+    ///
+    /// Returns an error when there already is an extension with the same extension type.
+    pub fn add(&mut self, extension: Extension) -> Result<(), InvalidExtensionError> {
+        if !self.contains(extension.extension_type()) {
+            self.inner.push(extension);
+            Ok(())
+        } else {
+            Err(InvalidExtensionError::Duplicate)
+        }
+    }
+
+    /// Add an extension to the extension list (or replace an existing one.)
+    ///
+    /// Returns the replaced extension (if any).
+    pub fn add_or_replace(&mut self, extension: Extension) -> Option<Extension> {
+        let replaced = self.remove(extension.extension_type()).ok();
+        self.inner.push(extension);
+        replaced
+    }
+
+    /// Remove an extension from the extension list.
+    ///
+    /// Returns the removed extension or an error when there is no extension with the given extension type.
+    pub fn remove(
+        &mut self,
+        extension_type: ExtensionType,
+    ) -> Result<Extension, InvalidExtensionError> {
```

Comment:
Data structures usually don't error when removin a non-existing element but return an `Option` and `None` in case the element is not removed. We should do the same here.